libs/parser/parser.c

```diff
@@ -0,0 +1,425 @@
+/*
+ * Phoenix-Pilot
+ *
+ * Parser for config files
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Piotr Nieciecki
+ *
+ * This file is part of Phoenix-Pilot software
+ *
+ * %LICENSE%
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <stdint.h>
+
+#include "parser.h"
+
+
+#define WORD "[A-Za-z0-9_.,-]"
+
+
+/*
+ * Patterns for regex expressions
+ *
+ * If you have changed these patterns check if these function work properly:
+ * - parser_headerGet
+ * - parser_fieldGet
+ *
+ * Especially be aware of `regmatch_t` structures.
+ */
+#define HEADER_PATTERN        "^[[:space:]]*@(" WORD "+)[[:space:]]*(#.*)?[[:space:]]*$"
+#define HEADER_SUBEXPRESSIONS 2
+
+#define FIELD_PATTERN        "^[[:space:]]*(" WORD "+)[ =]+(" WORD "+)([[:space:]]+#.*)?[[:space:]]*$"
+#define FIELD_SUBEXPRESSIONS 3
+
+#define COMMENT_PATTERN "^[[:space:]]*#.*$|^[[:space:]]*$"
+
+
+/* Static variables */
+struct {
+	int regexCompErr;
+
+	regex_t headerRegex;
+	regex_t fieldRegex;
+	regex_t commentRegex;
+} parser_common;
+
+
+/* Structure contains information about defined header */
+typedef struct {
+	char headerName[MAX_HEADER_LEN + 1];
+	int (*converter)(const hmap_t *);
+} parser_headerInfo_t;
+
+
+/* Contains information about particular field from file */
+typedef struct {
+	char field[MAX_FIELD_LEN + 1];
+	char value[MAX_VALUE_LEN + 1];
+} parser_field_t;
+
+
+/* Structure of the main parser object */
+struct parser_t {
+	parser_headerInfo_t *headerInfos;
+	hmap_t *headersMap;
+
+	parser_field_t *fields;
+	hmap_t *fieldsMap;
+};
+
+
+enum parser_lineType { type_Header, type_Field, type_Comment, type_InvalidLine };
+
+
+parser_t *parser_alloc(int maxHeadersNb, int maxFieldsNb)
+{
+	parser_t *result;
+
+	if (parser_common.regexCompErr != 0) {
+		fprintf(stderr, "Parser: an error occurred during Regex compilation.\n");
+		return NULL;
+	}
+
+	if (maxHeadersNb <= 0 || maxFieldsNb <= 0) {
+		fprintf(stderr, "Parser: invalid arguments. Both of them have to be bigger than zero.\n");
```

Comment:
Unclear error message. User may be confused about _both of  what_ has to be bigger than zero. You can use variable names directly like:
```suggestion
		fprintf(stderr, "Parser: invalid maxHeadersNb/maxFieldsNb arguments\n");
```
Or change those arguments to unsigned int and then have only:
```suggestion
		fprintf(stderr, "Parser: invalid arguments. Header or fields number equal to zero.\n");
```